pkg/query/books.go

```diff
@@ -0,0 +1,126 @@
+package query
+
+import "book_service/pkg/constants"
+
+type Builder struct {
+	id           *string
+	title        *string
+	authorName   *string
+	priceMin     *float64
+	priceMax     *float64
+	aggregations map[string]interface{}
+}
+
+func NewQueryBuilder() *Builder {
+	return &Builder{
+		aggregations: make(map[string]interface{}),
+	}
+}
+
+func (qb *Builder) ID(i string) *Builder {
+	if i == "" {
+		return qb
+	}
+	qb.id = &i
+	return qb
+}
+
+func (qb *Builder) Title(t string) *Builder {
+	if t == "" {
+		return qb
+	}
+	qb.title = &t
+	return qb
+}
+
+func (qb *Builder) AuthorName(a string) *Builder {
+	if a == "" {
+		return qb
+	}
+	qb.authorName = &a
+	return qb
+}
+
+func (qb *Builder) PriceRange(min, max float64) *Builder {
+	unseted := min == max && min == 0
+	if unseted {
+		qb.priceMin = &min
+		highestPrice := constants.HighestBookPrice
+		qb.priceMax = &highestPrice
+		return qb
+	}
+	qb.priceMin = &min
+	qb.priceMax = &max
+	return qb
+}
+
+func (qb *Builder) DistinctAuthors() *Builder {
+	qb.aggregations["distinct_authors"] = map[string]interface{}{
+		"cardinality": map[string]interface{}{
+			"field": "author_name",
+		},
+	}
+	return qb
+}
+
+func (qb *Builder) TotalBooks() *Builder {
+	qb.aggregations["total_books"] = map[string]interface{}{
+		"value_count": map[string]interface{}{
+			"field": "_id",
+		},
+	}
+	return qb
+}
+
+func (qb *Builder) Build() map[string]interface{} {
+	var mustClauses []map[string]interface{}
+
+	if qb.id != nil {
+		mustClauses = append(mustClauses, map[string]interface{}{
+			"term": map[string]interface{}{
+				"_id": *qb.id,
+			},
+		})
+	}
+
+	if qb.title != nil {
+		mustClauses = append(mustClauses, map[string]interface{}{
+			"match": map[string]interface{}{
+				"title": *qb.title,
+			},
+		})
+	}
+
+	if qb.authorName != nil {
+		mustClauses = append(mustClauses, map[string]interface{}{
+			"match": map[string]interface{}{
+				"author_name": *qb.authorName,
+			},
+		})
+	}
+
+	if qb.priceMin != nil && qb.priceMax != nil {
```

Comment:
What happens if only min price was given?
I don't see a reason to limit the max price in the query...

Reply:
In the price range builder we are making sure it can't happen